src/mscorlib/src/System/Collections/Generic/List.cs

```diff
@@ -247,7 +246,7 @@ public void Add(T item)
             var array = _items;
             var size = _size;
             _version++;
-            if ((uint)size < (uint)array.Length)
+            if (array.IndexInRange(size))
```

Comment:
>  Can we fix the JIT to morph this code into unsigned comparison instead of doing it manually?

I suspect that it may be possible in certain situations but in the general case I'm afraid that the answer is no. Let's look at the example that started all this - `List<T>.Add(T)`:
```C#
if (size < array.Length)
{
    _size = size + 1;
    array[size] = item;
 }
 else
 {
     AddWithResize(item);
}
```
The JIT "sees" `size < array.Length` but this is not enough to eliminate the array range check because `size` might be negative. Nor can the JIT change `size < array.Length` to unsigned, doing so would change the meaning of the program. If `size` happens to be negative then the original code would throw an exception while the new code would execute the `AddWithResize` branch.

By using `(uint)size < (uint)array.Length` the developer is adding information that the JIT simply doesn't have and helps it eliminate the range check.

It may be possible for the JIT to recognize the more intuitive `(0 <= size && size < array.Length)` but it will have to work harder for that.

That said, I like clean code and I think that such tricks should only be used when really needed. And I've played enough with this kind of stuff to know that saving a few range checks here and there won't necessarily make a difference for real word applications.